### ISSUE TITLE
FDSN client: problem with redirection from http to https 

### ISSUE THREAD

**Issue by SCEDC** [obspy/obspy#1665]
Hi,

obspy.clients.fdsn.Client does not seem to work when the service endpoint is forwarded to https. We've been transitioning the SCEDC web services to use https. Once we set http to forward to https, ObsPy requests started getting a bad request error. 

This is the sample code:

```
from obspy import UTCDateTime
from obspy.clients.fdsn import Client

client = Client("SCEDC")
catalog = client.get_events(minmagnitude=5.0, maxmagnitude=6.0, starttime="2010-01-01")
print(catalog)
```

The output when http to https forwarding is enabled on our server is:

```
Traceback (most recent call last):
  File "./obspy_example.py", line 7, in <module>
    catalog = client.get_events(minmagnitude=8.0, starttime="2010-01-01")
  File "/export/nobackup/schen/anaconda3/lib/python3.5/site-packages/obspy/clients/fdsn/client.py", line 415, in get_events
    data_stream = self._download(url)
  File "/export/nobackup/schen/anaconda3/lib/python3.5/site-packages/obspy/clients/fdsn/client.py", line 1342, in _download
    raise FDSNException(msg, server_info)
obspy.clients.fdsn.header.FDSNException: Bad request. If you think your request was valid please contact the developers.

```
When forwarding is disabled, the same code is able to retrieve the catalog:

```
1008 Event(s) in Catalog:
2010-05-02T14:52:43.400000Z | -34.212,  -71.864 | 5.9 Mw | automatic
2010-04-05T00:07:07.920000Z | +31.979, -115.134 | 5.22 Ml | manual
...
2013-07-15T13:59:04.500000Z | -15.434, -173.058 | 5.8 Mw | automatic
2014-11-23T11:01:25.300000Z | +43.798, -128.408 | 5.4 Mb | automatic
To see all events call 'print(CatalogObject.__str__(print_all=True))'
```

Right now, we've disabled http to https forwarding to allow users to access our web services through ObsPy. I'm  using ObsPy 1.0.2 installed through Anaconda to test this issue. 

Shang-Lin Chen

**Comment by megies** [comment]
One thing you can always do is explicitly specify the URL, e.g. `client = Client("https://service.scedc.caltech.edu")`, but I understand that it's not feasible to convey to all users..

Also, this sounds a bit like  #1143..

**Comment by markcwill** [comment]
Not enough info to troubleshoot this, and can't test on your site since you've disabled the feature. Did this request using a URL work with `requests` and/or a recent web browser?

What exactly do you mean by "forwarding"? Are you sending a 301 or 303 redirect? If so, then this might be related to #1143 as @megies said. If you are routing, rewriting requests, proxying or port forwarding on your end, then it may be an issue with the service. Based on your traceback, I think you're hitting this condition:
https://github.com/obspy/obspy/blob/1.0.2/obspy/clients/fdsn/client.py#L1339

So your server is returning a 400, which means this bad request could be coming from the obspy client, or it could be coming from your "forwarding" service. If you have the logs of the bad request, that may help as well. Can you tell if the original redirect happened, and this is an error for the subsequent request, or is this error happening on the initial "http" request?

p.s. @megies, is `requests` a required package now?

**Comment by megies** [comment]
> p.s. @megies, is requests a required package now?

It has been a dependency for a long while, the whole 1.0.x series (see setup.py; it was added with #1227 and we started to transition more and more network interaction to it for more concise code).

**Comment by SCEDC** [comment]
When forwarding was active, requests in the browser and with wget 1.12 worked. Requests from an older version of wget reportedly did not work. I didn't try `requests`.

**Comment by megies** [comment]
@SCEDC I guess this will be hard (if not impossible) to debug unless we have a test service.. would it be possible for you to set up a http/https-forward service just for testing somewhere?

**Comment by SCEDC** [comment]
Our other web server, robson.gps.caltech.edu now has http/https forwarding. wget and requests work. 

```
In [6]: r = requests.get('https://robson.gps.caltech.edu/fdsnws/station/1/query?nodata=404&net=CI&sta=PASC&cha=HHE')

In [7]: r.status_code
Out[7]: 200

```

ObsPy only works when I set the client to https://robson.gps.caltech.edu.

```
In [1]: from obspy.clients.fdsn import Client

In [2]: client = Client("http://robson.gps.caltech.edu")

In [3]: catalog = client.get_events(minmagnitude=5.0, maxmagnitude=6.0, starttim
   ...: e="2010-01-01")
---------------------------------------------------------------------------
FDSNException                             Traceback (most recent call last)
<ipython-input-3-9930214febb4> in <module>()
----> 1 catalog = client.get_events(minmagnitude=5.0, maxmagnitude=6.0, starttime="2010-01-01")

/export/nobackup/schen/anaconda3/lib/python3.5/site-packages/obspy/clients/fdsn/client.py in get_events(self, starttime, endtime, minlatitude, maxlatitude, minlongitude, maxlongitude, latitude, longitude, minradius, maxradius, mindepth, maxdepth, minmagnitude, maxmagnitude, magnitudetype, includeallorigins, includeallmagnitudes, includearrivals, eventid, limit, offset, orderby, catalog, contributor, updatedafter, filename, **kwargs)
    413             "event", DEFAULT_PARAMETERS['event'], kwargs)
    414 
--> 415         data_stream = self._download(url)
    416         data_stream.seek(0, 0)
    417         if filename:

/export/nobackup/schen/anaconda3/lib/python3.5/site-packages/obspy/clients/fdsn/client.py in _download(self, url, return_string, data, use_gzip)
   1340             msg = ("Bad request. If you think your request was valid "
   1341                    "please contact the developers.")
-> 1342             raise FDSNException(msg, server_info)
   1343         elif code == 401:
   1344             raise FDSNException("Unauthorized, authentication required.",

FDSNException: Bad request. If you think your request was valid please contact the developers.

In [4]: client = Client("https://robson.gps.caltech.edu")

In [5]: catalog = client.get_events(minmagnitude=5.0, maxmagnitude=6.0, starttim
   ...: e="2010-01-01")

In [6]: print(catalog)
1007 Event(s) in Catalog:
2010-05-02T14:52:43.400000Z | -34.212,  -71.864 | 5.9 Mw | automatic
2010-04-05T00:07:07.920000Z | +31.979, -115.134 | 5.22 Ml | manual
...
2016-09-10T10:08:19.600000Z |  -5.583,  -76.965 | 6.0 Mb | automatic
2016-12-21T16:43:58.330000Z | +21.528, +145.382 | 5.9 Mun | automatic
To see all events call 'print(CatalogObject.__str__(print_all=True))'

```

Setting force_redirect to True for http://robson.gps.caltech.edu does not work.

```
In [7]: client = Client("http://robson.gps.caltech.edu", force_redirect=True)

In [8]: catalog = client.get_events(minmagnitude=5.0, maxmagnitude=6.0, starttim
   ...: e="2010-01-01")
---------------------------------------------------------------------------
FDSNException                             Traceback (most recent call last)
<ipython-input-8-9930214febb4> in <module>()
----> 1 catalog = client.get_events(minmagnitude=5.0, maxmagnitude=6.0, starttime="2010-01-01")

/export/nobackup/schen/anaconda3/lib/python3.5/site-packages/obspy/clients/fdsn/client.py in get_events(self, starttime, endtime, minlatitude, maxlatitude, minlongitude, maxlongitude, latitude, longitude, minradius, maxradius, mindepth, maxdepth, minmagnitude, maxmagnitude, magnitudetype, includeallorigins, includeallmagnitudes, includearrivals, eventid, limit, offset, orderby, catalog, contributor, updatedafter, filename, **kwargs)
    413             "event", DEFAULT_PARAMETERS['event'], kwargs)
    414 
--> 415         data_stream = self._download(url)
    416         data_stream.seek(0, 0)
    417         if filename:

/export/nobackup/schen/anaconda3/lib/python3.5/site-packages/obspy/clients/fdsn/client.py in _download(self, url, return_string, data, use_gzip)
   1340             msg = ("Bad request. If you think your request was valid "
   1341                    "please contact the developers.")
-> 1342             raise FDSNException(msg, server_info)
   1343         elif code == 401:
   1344             raise FDSNException("Unauthorized, authentication required.",

FDSNException: Bad request. If you think your request was valid please contact the developers.

```

**Comment by markcwill** [comment]
Thanks @SCEDC! When I use requests it works fine, and the 301 redirect req is in the history. Everything seems to be in order. When I run your query with obspy, I get this:
```
FDSNException: Bad request. If you think your request was valid please contact the developers.
Detailed response of server:

Error 400: Bad date value for starttime: 2010-01-01T00%3A00%3A00.000000
Request:
https://robson.gps.caltech.edu/fdsnws/event/1/query?minmagnitude=5.0&maxmagnitude=6.0&starttime=2010-01-01T00%253A00%253A00.000000
Request Submitted:
2017/02/17 02:43:48 UTC
Service version:
fdsnws-event: v 1.1.1
```
The "Bad date value" actually looks ok, but seems that the redirect is causing a double pass of urlencode, which is trying to encode the encodings?
```python
In [16]: urllib.parse.urlencode({'time':'2010-01-01T00:00:00.000000'})
Out[16]: 'time=2010-01-01T00%3A00%3A00.000000'

In [17]: urllib.parse.urlencode({'time':'2010-01-01T00%3A00%3A00.000000'})
Out[17]: 'time=2010-01-01T00%253A00%253A00.000000'
```
@megies, this is where I lose the trail, is the custom `HTTPRedirectHandler` calling `urlencode` again under the hood? Long term, would be cool to use `requests`, which I think handles `POST` properly for 303s, etc, etc. For now I think calling `urllib.parse.unquote` on the `newurl` in `CustomRedirectHandler` will fix? I can do a PR tomorrow...

**Comment by megies** [comment]
> Long term, would be cool to use requests

Yeah, we figured that too.. see #1254 but nobody had time to do it so far..

> I can do a PR tomorrow...

Would be great! :-)

**Comment by markcwill** [comment]
@SCEDC, I believe the problem is with your redirect service on port 80.

So, `requests` DOES NOT work properly with your redirect service (and neither does a browser), or rather, only works when we skirt around RFC 2616 and send colons in the query string without encoding them. I think your redirect service is correctly trying to encode the URL query params in the `Location` header, but you are blanket encoding whatever is already there. 

The obspy `fdsn.Client` (and `requests`, when asked to do so) is correctly encoding the colons in `starttime`, but your redirect service is not unencoding or unquoting the query params before running an additional encoder on the URL. Therefore, you are encoding the `%` sign of the already-encoded `:` characters in the query arguments which contain times.

Here are some examples that have nothing to do with obspy:

This returns an incorrect URI in the `Location` header which causes your web service to return an error
```python
In [58]: requests.get("http://robson.gps.caltech.edu/fdsnws/event/1/query", params={'starttime':'2016-01-01T00:00:00.000000'}, allow_redirects=False).headers.get('Location')
Out[58]: 'https://robson.gps.caltech.edu/fdsnws/event/1/query?starttime=2016-01-01T00%253A00%253A00.000000'
```
You will get the same error if you get this URL in your browser:
http://robson.gps.caltech.edu/fdsnws/event/1/query?minmagnitude=5.0&maxmagnitude=6.0&starttime=2016-01-01T00%3A00%3A00.000000

Unsolicited advice: I guess you can do one of two things with your redirect service: 
1. Don't bother re-encoding the query string arguments. This might not be 100% RFC-compliant, but it will solve all your problems. If you are just trying to redirect from HTTP to HTTPS, it won't really affect you -- it appears your FDSN web service accepts encoded OR unencoded colons in the args, so this would be just passing the buck to your "real" service on 443, and matches the FDSN spec:

> Requests should be URL Encoded as a standard practice before submission to the service, but services should handle non-encoded submissions as far as is possible.	

2. Parse and unquote the query string params and re-encode them, either systematically, or in a custom way. This would add a lot of complexity and/or heft to your service, but would be RFC 2616 compliant wrt header field formatting.

**Comment by megies** [comment]
Punting this to 1.1.0, sounds like it's potentially an external problem anyway..

**Comment by markcwill** [comment]
I think we can close, was leaving open until we hear from Caltech, pinging @SCEDC?

**Comment by megies** [comment]
Ok, sounds like this is an external problem according to very detailed analysis of @markcwill (thanks!).

Closing. @SCEDC feel free to reopen if you come up with a different conclusion and feel like we need to change something on our end.

**Comment by markcwill** [comment]
Same problem as #1048 also affecting the GS in #1690

**Comment by krischer** [comment]
Should be "fixed" in #1690 by just making the mapping use HTTPS.

**Comment by SCEDC** [comment]
The issue was resolved on our end by adding NE to the end of the RewriteRule in our Apache vhosts config.
